protocol.go

```diff
@@ -48,6 +48,45 @@ const (
 
 var errNoTimeout = errors.New("no timeout")
 
+// ProtocolFromRequest returns the inferred protocol name for parsing an
+// HTTP request. It inspects the request's method and headers to determine the
+// protocol. If the request doesn't match any known protocol, an empty string
+// is returned.
+func ProtocolFromRequest(request *http.Request) (string, bool) {
+	switch classifyRequest(request, false) {
+	case connectUnaryProtocol, connectStreamProtocol:
+		return ProtocolConnect, true
+	case grpcProtocol:
+		return ProtocolGRPC, true
+	case grpcWebProtocol:
+		return ProtocolGRPCWeb, true
+	case unknownProtocol:
+		return "", false
+	default:
+		return "", false
+	}
+}
+
+// ProcedureFromURL returns the inferred procedure name from a URL. It's
+// returned in the form "/service/method" if a valid suffix is found. If the
+// path doesn't contain a service and method, the entire path is returned.
+func ProcedureFromURL(url *url.URL) (string, bool) {
```

Comment:
Why isn't this method used anywhere? The code that extracts this to report in the `Spec` should use this, no?

Reply:
The handler returns the procedure name to be passed to a router and the generated code can then hardcode the method name.